### PR TITLE
Tighten up shell script

### PR DESCRIPTION
**Pull request by ddgenome**
Creating a PR so I can comment on the parts of the script I think are about enhancing the capabilities of container skills.

### CODE REVIEW

File: markdownlint.sh [hunk 1]
```diff
@@ -1,64 +1,113 @@
...
 
-if ! [[ -z "ignores" ]]; then
-  ignore_option="--ignore $ignores"
-fi
+    local outdir=${ATOMIST_OUTPUT_DIR:-/atm/output}
 
-markdownlint "**/*.md" $config_option $ignore_option $fix_option
+    # Make the problem matcher available to the runtime
+    local matcher_outdir=$outdir/matchers
+    if ! mkdir -p "$matcher_outdir"; then
```
Comment (cdupuis) [inline markdownlint.sh:57]: That directory will always exist. The runtime creates it. Is that going to cause an issue?

File: markdownlint.sh [hunk 2]
```diff
@@ -1,64 +1,113 @@
...
+        err "Failed to create matcher output directory: $matcher_outdir"
+        return 1
+    fi
+    if ! cp /app/markdownlint.matcher.json "$matcher_outdir"; then
```
Comment (ddgenome) [inline markdownlint.sh:61]: Is the matcher stuff the same as the problem extraction stuff you can do on the container skill configuration? It's nice to have this flexibility, but out of box if the container skill is running on a push, it should set a check based on the exit value.

Is there some magic to the name `markdownlint.matcher.json`? Is a glob done on `/app/*.matcher.json`?

File: markdownlint.sh [hunk 3]
```diff
@@ -1,64 +1,113 @@
...
+    fi
+
+    # Create push instructions for the runtime to indicate how changes to the repo should get persisted
+    if [[ ! $labels ]]; then
+        labels="[]"
+    fi
+
+    if ! > "$outdir/push.json" jq -n --arg s "$fix" --argjson l "$labels" '{
+    strategy: $s,
+    pullRequest: {
```
Comment (ddgenome) [inline markdownlint.sh:73]: It seems we have settled on snake_case everywhere else. Why not here?

File: markdownlint.sh [hunk 4]
```diff
@@ -1,64 +1,113 @@
...
-if ! [[ -z "$fix" ]]; then
-  fix_option="--fix"
-fi
+function main () {
+    # Extract some skill configuration from the incoming event payload
+    local fix config ignores labels branch
+    fix=$( < "$ATOMIST_PAYLOAD" \
+               jq -r '.skill.configuration.instances[0].parameters[] | select( .name == "push_strategy" ) | .value' )
+    config=$( < "$ATOMIST_PAYLOAD" \
+                  jq -r '.skill.configuration.instances[0].parameters[] | select( .name == "config" ) | .value' )
```
Comment (ddgenome) [inline markdownlint.sh:41]: Should this use `-r` if the config value is JSON? I guess it does not matter as `-r` really only affects strings.

File: markdownlint.sh [hunk 5]
```diff
@@ -1,64 +1,113 @@
...
+    ignores=$( < "$ATOMIST_PAYLOAD" \
+                   jq -r '.skill.configuration.instances[0].parameters[] | select( .name == "ignores" ) | .value | join(" --ignore ")' )
+    labels=$( < "$ATOMIST_PAYLOAD" \
+                  jq -r '.skill.configuration.instances[0].parameters[] | select( .name == "labels" ) | .value' )
```
Comment (ddgenome) [inline markdownlint.sh:45]: Same question about `-r`.

File: markdownlint.sh [hunk 6]
```diff
@@ -1,64 +1,113 @@
...
+    if [[ ! $labels ]]; then
+        labels="[]"
+    fi
+
+    if ! > "$outdir/push.json" jq -n --arg s "$fix" --argjson l "$labels" '{
+    strategy: $s,
+    pullRequest: {
+      title: "MarkdownLint fixes",
+      body: "MarkdownLint fixed warnings and/or errors",
+      branchPrefix: "markdownlint",
```
Comment (ddgenome) [inline markdownlint.sh:76]: snake_case

File: markdownlint.sh [hunk 7]
```diff
@@ -1,64 +1,113 @@
...
+        err "Failed to copy markdownlink.matcher.json to $matcher_outdir"
+        return 1
+    fi
+
+    # Create push instructions for the runtime to indicate how changes to the repo should get persisted
+    if [[ ! $labels ]]; then
+        labels="[]"
+    fi
+
+    if ! > "$outdir/push.json" jq -n --arg s "$fix" --argjson l "$labels" '{
```
Comment (ddgenome) [inline markdownlint.sh:71]: Is the path `$ATOMIST_OUTPUT_DIR/push.json` provided as an environment variable or just a convention?

File: markdownlint.sh [hunk 8]
```diff
@@ -1,64 +1,113 @@
...
+      labels: $l,
+      close: {
+        stale: true,
+        message: "Closing pull request because all fixable warnings and/or errors have been fixed in base branch"
+      }
+    },
+    commit: {
+      message: "MarkdownLint fixes"
+    }
+}'
```
Comment (ddgenome) [inline markdownlint.sh:86]: Like the problem extraction, tt is nice to have the power provided by the `push.json`, but I'd be happy with just a check based on the exit status.

File: markdownlint.sh [hunk 9]
```diff
@@ -1,64 +1,113 @@
...
+        local config_file=$inputdir/markdownlint.config.json
+        echo "$config" > "$config_file"
+        config_option="--config $config_file"
+    fi
+    local ignore_option
+    if [[ $ignores ]]; then
+        ignore_option="--ignore $ignores"
+    fi
+
+    exec markdownlint "**/*.md" $config_option $ignore_option $fix_option
```
Comment (ddgenome) [inline markdownlint.sh:110]: I guess the `markdownlint` executable does its own glob expansion?

File: markdownlint.sh (hunk 1 above)
Comment (ddgenome) [reply markdownlint.sh:57]:
```man
     -p      Create intermediate directories as required.  If this option is
             not specified, the full path prefix of each operand must already
             exist.  On the other hand, with this option specified, no error
             will be reported if a directory given as an operand already
             exists.  Intermediate directories are created with permission
             bits of rwxrwxrwx (0777) as modified by the current umask, plus
             write and search permission for the owner.
```

If it always exists, is it provided as an environment variable or just a convention?

File: markdownlint.sh (hunk 8 above)
Comment (cdupuis) [reply markdownlint.sh:86]: That’s already an option. You can get just a check.

File: markdownlint.sh (hunk 2 above)
Comment (cdupuis) [reply markdownlint.sh:61]: Yes, that’s the same stuff as the container-run config. You can set a check purely on the exit code. 

No, the name doesn’t matter. Just that it has to be inside the marchers folder

File: markdownlint.sh (hunk 3 above)
Comment (cdupuis) [reply markdownlint.sh:73]: I don’t think we settled on snake case. Where do we use that in the skill world? We used to in atomist.yaml but that’s dead.

File: markdownlint.sh (hunk 9 above)
Comment (cdupuis) [reply markdownlint.sh:110]: It does.

File: markdownlint.sh (hunk 3 above)
Comment (ddgenome) [reply markdownlint.sh:73]: atomist.yaml is what I was thinking of.

File: markdownlint.sh (hunk 7 above)
Comment (cdupuis) [reply markdownlint.sh:71]: Right now just a convention.

**Pull request merged**